packages/ymax-planner/src/cosmos-rest-client.ts

```diff
@@ -0,0 +1,252 @@
+
+    console.warn(
+      `[CosmosRestClient] Fetching ${denom} balance for ${address} on ${chainConfig.name}: ${url}`,
+    );
+
+    const response = await this.makeRequest<{ balance: Coin }>(
+      url,
+      chainConfig,
+      `${denom} balance for ${address} on ${chainConfig.name}`,
+    );
+
+    return response.balance;
+  }
+
+  /**
+   * Get chain information/status
+   */
+  async getChainInfo(chainKey: string) {
+    const chainConfig = this.chainConfigs.get(chainKey);
+    if (!chainConfig) {
+      throw new Error(`Chain configuration not found for: ${chainKey}`);
+    }
+
+    const url = `${chainConfig.restEndpoint}/cosmos/base/tendermint/v1beta1/node_info`;
+
+    console.warn(
+      `[CosmosRestClient] Fetching chain info for ${chainConfig.name}: ${url}`,
+    );
+
+    return this.makeRequest(
+      url,
+      chainConfig,
+      `Chain info for ${chainConfig.name}`,
+    );
+  }
+
+  private async makeRequest<T>(
+    url: string,
+    chainConfig: ChainConfig,
+    context: string,
+  ): Promise<T> {
+    await null;
+
+    let lastError: Error | null = null;
+
+    for (let attempt = 1; attempt <= this.retries; attempt += 1) {
+      try {
+        const controller = new AbortController();
+        const timeoutId = setTimeout(() => controller.abort(), this.timeout);
+
+        const response = await fetch(url, {
+          signal: controller.signal,
+          headers: {
+            Accept: 'application/json',
+            'User-Agent': 'Agoric-Portfolio-Planner/1.0.0',
+          },
+        });
+
+        clearTimeout(timeoutId);
+
+        if (!response.ok) {
+          const errorBody = await response.text().catch(() => 'Unknown error');
+          const error = new Error(
+            `HTTP ${response.status}: ${response.statusText} - ${errorBody}`,
+          ) as CosmosApiError;
+          error.statusCode = response.status;
+          error.chainId = chainConfig.chainId;
+          error.endpoint = url;
+          throw error;
+        }
+
+        const data = await response.json();
+        console.warn(`[CosmosRestClient] Success: ${context}`);
+        return data;
+      } catch (error) {
+        lastError = error instanceof Error ? error : new Error(String(error));
+
+        if (attempt < this.retries) {
+          const delay = Math.min(1000 * 2 ** (attempt - 1), 5000); // Exponential backoff, max 5s
+          console.warn(
+            `[CosmosRestClient] Attempt ${attempt} failed for ${context}, retrying in ${delay}ms:`,
+            lastError.message,
+          );
+          await new Promise(resolve => setTimeout(resolve, delay));
+        }
+      }
+    }
+
+    const finalError = new Error(
+      `Failed to fetch ${context} after ${this.retries} attempts: ${lastError?.message}`,
+    ) as CosmosApiError;
+    finalError.statusCode = 0;
+    finalError.chainId = chainConfig.chainId;
+    throw finalError;
```

Comment:
With error construction being specialized, we should consider abstraction.
```suggestion
    this.throwCosmosApiError(
      `Failed to fetch ${context} after ${this.retries} attempts: ${lastError?.message}`,
      { chainId: chainConfig.chainId, url },
    );
```

```ts
  private throwCosmosApiError(
    message: string,
    {
      statusCode = 0,
      chainId,
      url,
    }: { statusCode?: number; chainId: string; url: string },
  ): never {
    const err = new Error(message) as CosmosApiError;
    err.statusCode = statusCode;
    err.chainId = chainId;
    error.endpoint = url;
    throw error;
  }
```